File: src/kotoba/lang/kami_nv_compat/controllers/operational_space.clj
(ns kotoba.lang.kami-nv-compat.controllers.operational-space
  "OperationalSpaceController — JVM port of src/controllers/operational-space.ts.
  Task-space torque control (sibling of DifferentialIKController): target
  pose/wrench → joint TORQUE via Jacobian transpose. Cartesian impedance
  (K_p·pos_err + K_d·vel_err) → 6-DOF task wrench → τ = Jᵀ F_task, with
  optional null-space joint regularization (n>6) + gravity compensation.
  Algorithm-for-algorithm port. Pure, zero runtime deps (reuses the
  differential-ik quaternion math). Wave 11 of ADR-2607020130."
  (:require [kotoba.lang.kami-nv-compat.controllers.differential-ik :as dik]))

;; ── Config ────────────────────────────────────────────────────────────────

(defn make-default-osc-cfg
  ([] (make-default-osc-cfg nil))
  ([overrides]
   (merge {:target-types              ["pose_abs"]
           :impedance-mode            "fixed"
           :motion-stiffness-task     [100 100 100 100 100 100]
           :motion-damping-ratio-task [1 1 1 1 1 1]
           :motion-stiffness-limits   [0 1000]
           :motion-damping-limits     [0 100]
           :nullspace-control         "none"
           :nullspace-stiffness       10
           :nullspace-damping-ratio   1
           :gravity-compensation      false}
          overrides)))

;; ── Null-space projection (damped pseudoinverse, Gauss-Jordan 6×6) ────────

(defn- project-to-nullspace
  "Project tauNs into the null-space of J: P_null = I - Jᵀ(JJᵀ+λ²I)⁻¹ J."
  [J tauNs n lam]
  (let [lam2 (* lam lam)
        ;; Jtau = J · tauNs (6-vec)
        Jtau (double-array (for [k (range 6)]
                             (loop [i 0 acc 0.0]
                               (if (>= i n) acc
                                 (recur (inc i) (+ acc (* (double (get-in J [k i])) (double (nth tauNs i)))))))))
        ^"[[D" aug (make-array Double/TYPE 6 7)]
    ;; A = J Jᵀ + λ²I, augmented with Jtau.
    (dotimes [i 6]
      (dotimes [j 6]
        (let [s (loop [k 0 acc 0.0]
                  (if (>= k n) acc
                    (recur (inc k) (+ acc (* (double (get-in J [i k])) (double (get-in J [j k])))))))]
          (aset ^"[[D" aug i j (double (+ s (if (= i j) lam2 0))))))
      (aset ^"[[D" aug i 6 (aget Jtau i)))
    ;; Gauss-Jordan.
    (dotimes [col 6]
      (let [piv (loop [r (inc col) p col mx (Math/abs (aget ^"[[D" aug col col))]
                  (if (>= r 6) p
                    (let [v (Math/abs (aget ^"[[D" aug r col))]
                      (if (> v mx) (recur (inc r) r v) (recur (inc r) p mx)))))
            mx  (Math/abs (aget ^"[[D" aug piv col))]
        (when (>= mx 1e-18)
          (when (not= piv col)
            (let [tmp (aget ^"[[D" aug col)]
              (aset ^"[[D" aug col (aget ^"[[D" aug piv))
              (aset ^"[[D" aug piv tmp)))
          (let [pv (aget ^"[[D" aug col col)]
            (dotimes [j 7] (aset ^"[[D" aug col j (/ (aget ^"[[D" aug col j) pv))))
          (dotimes [r 6]
            (when (not= r col)
              (let [f (aget ^"[[D" aug r col)]
                (when (>= (Math/abs f) 1e-18)
                  (dotimes [j 7]
                    (aset ^"[[D" aug r j (- (aget ^"[[D" aug r j)
                                            (* f (aget ^"[[D" aug col j))))))))))))
    ;; Jty = Jᵀ y ; out = tauNs - Jty.
    (let [y (double-array (for [i (range 6)] (aget ^"[[D" aug i 6)))]
      (vec (for [i (range n)]
             (- (double (nth tauNs i))
                (loop [k 0 acc 0.0]
                  (if (>= k 6) acc
                    (recur (inc k) (+ acc (* (double (get-in J [k i])) (aget y k))))))))))))

;; ── Controller ────────────────────────────────────────────────────────────

(defprotocol IOperationalSpaceController
  (cfg        [this])
  (num-envs   [this])
  (num-dof    [this])
  (action-dim [this])
  (reset-controller! [this env-ids])
  (set-command! [this command opts])
  (get-target  [this env-idx])
  (compute     [this args]))

(defn- target-type-dim [tt]
  (condp = tt "pose_abs" 7 "pose_rel" 6 "wrench_abs" 6 3))   ; force_abs / torque_abs

(defn operational-space-controller
  "Build an OperationalSpaceController. `num-envs` default 1, `num-dof` 7."
  ([cfg] (operational-space-controller cfg 1 7))
  ([cfg num-envs] (operational-space-controller cfg num-envs 7))
  ([cfg num-envs num-dof]
   (when (<= num-envs 0) (throw (ex-info "num-envs must be > 0" {:num-envs num-envs})))
   (when (<= num-dof 0) (throw (ex-info "num-dof must be > 0" {:num-dof num-dof})))
   (when (not (#{"fixed" "variable"} (:impedance-mode cfg)))
     (throw (ex-info "impedance-mode must be 'fixed' or 'variable'" {:impedance-mode (:impedance-mode cfg)})))
   (when (not (#{"none" "position"} (:nullspace-control cfg)))
     (throw (ex-info "nullspace-control must be 'none' or 'position'" {:nullspace-control (:nullspace-control cfg)})))
   (when (not= 6 (count (:motion-stiffness-task cfg)))
     (throw (ex-info "motion-stiffness-task must be 6-vec" {:got (count (:motion-stiffness-task cfg))})))
   (when (not= 6 (count (:motion-damping-ratio-task cfg)))
     (throw (ex-info "motion-damping-ratio-task must be 6-vec" {:got (count (:motion-damping-ratio-task cfg))})))
   (let [target  (atom (vec (repeat num-envs [0 0 0 0 0 0 1])))
         var-stf (atom (vec (repeat num-envs nil)))]
     (reify IOperationalSpaceController
       (cfg      [_] cfg)
       (num-envs [_] num-envs)
       (num-dof  [_] num-dof)
       (action-dim [_]
         (let [base (reduce + (map target-type-dim (:target-types cfg)))]
           (if (= (:impedance-mode cfg) "variable") (+ base 12) base)))
       (reset-controller! [_ env-ids]
         (let [ids (or env-ids (range num-envs))]
           (swap! target (fn [t] (reduce #(assoc % %2 [0 0 0 0 0 0 1]) t ids)))
           (swap! var-stf (fn [v] (reduce #(assoc % %2 nil) v ids)))))
       (set-command! [_ command {:keys [ee-pos ee-quat env-idx] :or {env-idx 0}}]
         (when (or (neg? env-idx) (>= env-idx num-envs))
           (throw (ex-info (str "env-idx=" env-idx " out of range [0, " num-envs ")")
                           {:env-idx env-idx :num-envs num-envs})))
         (let [expected (if (= (:impedance-mode cfg) "variable")
                          (+ (reduce + (map target-type-dim (:target-types cfg))) 12)
                          (reduce + (map target-type-dim (:target-types cfg))))]
           (when (not= (count command) expected)
             (throw (ex-info (str "command must be length " expected) {:got (count command) :expected expected})))
           (let [[cmd _] (if (= (:impedance-mode cfg) "variable")
                           (let [gains-start (- expected 12)]
                             (swap! var-stf assoc env-idx (subvec command gains-start))
                             [(subvec command 0 gains-start)])
                           [command])]
             (loop [tts (:target-types cfg) idx 0]
               (when (seq tts)
                 (let [tt (first tts)]
                   (condp = tt
                     "pose_abs"
                     (do (swap! target assoc env-idx (subvec cmd idx (+ idx 7)))
                         (recur (rest tts) (+ idx 7)))
                     "pose_rel"
                     (do (when (or (nil? ee-pos) (nil? ee-quat))
                           (throw (ex-info "pose_rel requires ee-pos + ee-quat" {})))
                         (let [delta (subvec cmd idx (+ idx 6))
                               pos'  (vec (map + ee-pos (subvec delta 0 3)))
                               angle (Math/sqrt (reduce + (map #(* % %) (subvec delta 3 6))))
                               q'    (if (< angle 1e-9)
                                       ee-quat
                                       (let [ax (mapv #(/ % angle) (subvec delta 3 6))
                                             h  (* angle 0.5)
                                             s  (Math/sin h)]
                                         (dik/quat-mul [(* (ax 0) s) (* (ax 1) s) (* (ax 2) s) (Math/cos h)]
                                                       ee-quat)))]
                           (swap! target assoc env-idx (into pos' q'))
                           (recur (rest tts) (+ idx 6))))
                     "wrench_abs" (recur (rest tts) (+ idx 6))
                     (recur (rest tts) (+ idx 3)))))))))   ; force_abs / torque_abs
       (get-target [_ env-idx] (vec (@target (or env-idx 0))))
       (compute [_ {:keys [ee-pos ee-quat ee-lin-vel ee-ang-vel jacobian joint-pos joint-vel
                           gravity-torque nullspace-target-pos env-idx] :or {env-idx 0}}]
         (when (not= 6 (count jacobian))
           (throw (ex-info "jacobian must have 6 rows" {:rows (count jacobian)})))
         (let [n (count (nth jacobian 0))]
           (when (not= n num-dof)
             (throw (ex-info (str "jacobian width " n " != num-dof " num-dof) {:n n :num-dof num-dof})))
           (let [t (@target env-idx)
                 t-pos (subvec t 0 3) t-quat (subvec t 3 7)
                 pos-err (vec (map - t-pos ee-pos))
                 ori-err (dik/axis-angle-vec (dik/quat-mul t-quat (dik/quat-inverse ee-quat)))
                 error (vec (concat pos-err ori-err))
                 d-error (vec (concat (map - ee-lin-vel) (map - ee-ang-vel)))
                 vs (@var-stf env-idx)
                 [kp-task kd-ratio] (if (and (= (:impedance-mode cfg) "variable") vs)
                                      [(subvec vs 0 6) (subvec vs 6 12)]
                                      [(:motion-stiffness-task cfg) (:motion-damping-ratio-task cfg)])
                 kd-task (vec (for [i (range 6)] (* 2.0 (nth kd-ratio i) (Math/sqrt (max 0.0 (nth kp-task i))))))
                 f-task (vec (for [i (range 6)] (+ (* (nth kp-task i) (nth error i))
                                                   (* (nth kd-task i) (nth d-error i)))))
                 ;; τ = Jᵀ F_task
                 tau (vec (for [i (range n)]
                            (loop [k 0 acc 0.0]
                              (if (>= k 6) acc
                                (recur (inc k) (+ acc (* (double (get-in jacobian [k i])) (nth f-task k))))))))
                 ;; null-space regularization (n > 6)
                 tau (if (and (= (:nullspace-control cfg) "position")
                              nullspace-target-pos joint-pos joint-vel (> n 6))
                       (let [kp-ns (:nullspace-stiffness cfg)
                             kd-ns (* 2.0 (:nullspace-damping-ratio cfg) (Math/sqrt (max 0.0 kp-ns)))
                             tau-ns (vec (for [i (range n)]
                                           (- (* kp-ns (- (nth nullspace-target-pos i) (nth joint-pos i)))
                                              (* kd-ns (nth joint-vel i)))))
                             tau-ns-proj (project-to-nullspace jacobian tau-ns n 0.05)]
                         (vec (map + tau tau-ns-proj)))
                       tau)
                 ;; gravity compensation
                 tau (if (and (:gravity-compensation cfg) gravity-torque)
                       (let [m (min n (count gravity-torque))]
                         (vec (for [i (range n)] (+ (nth tau i) (if (< i m) (nth gravity-torque i) 0)))))
                       tau)]
             tau)))))))

File: src/controllers/operational-space.ts
// TypeScript port of kotodama.nv_compat.isaaclab.controllers.operational_space
//
// OperationalSpaceController — task-space torque control. Sibling of
// iter 72 DifferentialIKController:
//
//   - DifferentialIKController: target pose → joint POSITION delta via
//     DLS over Jacobian. Pairs with joint-position action.
//   - OperationalSpaceController: target pose / wrench → joint TORQUE
//     via Jacobian transpose. Pairs with joint-effort action directly.
//
// OSC = canonical contact-rich manipulation controller. Cartesian
// impedance (K_p · pos_err + K_d · vel_err) → 6-DOF task-space wrench →
// joint torque τ = Jᵀ F_task. Optional null-space joint regularization
// for redundant arms (n > 6) + optional gravity compensation.
//
// Algorithm-for-algorithm port of the Python iter 63 reference impl
// (~480 LoC Python).
//
// ADR-2605261800 §D6 nv-compat namespace localization. Pure TypeScript,
// zero runtime deps.

import { type Vec3, type Quat, quatMul, quatInverse, axisAngleVec } from "./differential-ik.js";

// ── Config ────────────────────────────────────────────────────────────────

export type OscTargetType =
  | "pose_abs"
  | "pose_rel"
  | "wrench_abs"
  | "force_abs"
  | "torque_abs";

export type OscImpedanceMode = "fixed" | "variable";

export type OscNullspaceControl = "none" | "position";

export interface OperationalSpaceControllerCfg {
  /** Target categories accepted by the controller (composed in order). */
  targetTypes: OscTargetType[];
  /** `"fixed"` — gains constant from cfg; `"variable"` — gains arrive in
   *  the command vector each step. */
  impedanceMode: OscImpedanceMode;
  /** 6-vec K_p_task (3 linear + 3 angular). */
  motionStiffnessTask: number[];
  /** 6-vec damping ratio ζ (1.0 = critically damped). */
  motionDampingRatioTask: number[];
  /** [min, max] gain bounds when impedanceMode="variable". */
  motionStiffnessLimits: [number, number];
  motionDampingLimits: [number, number];
  /** `"none"` or `"position"`. When `"position"`, regularize redundant
   *  DoFs toward `nullspaceTargetPos`. */
  nullspaceControl: OscNullspaceControl;
  /** Scalar gains for the null-space P/D loop. */
  nullspaceStiffness: number;
  nullspaceDampingRatio: number;
  /** When true, `compute` adds gravity-comp torque G(q) to the output. */
  gravityCompensation: boolean;
}

export function makeDefaultOscCfg(
  overrides: Partial<OperationalSpaceControllerCfg> = {},
): OperationalSpaceControllerCfg {
  return {
    targetTypes: ["pose_abs"],
    impedanceMode: "fixed",
    motionStiffnessTask: [100, 100, 100, 100, 100, 100],
    motionDampingRatioTask: [1, 1, 1, 1, 1, 1],
    motionStiffnessLimits: [0, 1000],
    motionDampingLimits: [0, 100],
    nullspaceControl: "none",
    nullspaceStiffness: 10,
    nullspaceDampingRatio: 1,
    gravityCompensation: false,
    ...overrides,
  };
}

// ── Controller ────────────────────────────────────────────────────────────

export class OperationalSpaceController {
  readonly cfg: OperationalSpaceControllerCfg;
  readonly numEnvs: number;
  readonly numDof: number;
  /** Per-env target pose [px, py, pz, qx, qy, qz, qw]. */
  private readonly _target: number[][];
  /** Per-env variable gains (length 12 = 6 stiffness + 6 damping ratio)
   *  or null when in fixed mode. */
  private readonly _variableStiffness: (number[] | null)[];

  constructor(
    cfg: OperationalSpaceControllerCfg,
    numEnvs: number = 1,
    numDof: number = 7,
  ) {
    if (numEnvs <= 0) throw new Error(`numEnvs must be > 0; got ${numEnvs}`);
    if (numDof <= 0) throw new Error(`numDof must be > 0; got ${numDof}`);
    if (cfg.impedanceMode !== "fixed" && cfg.impedanceMode !== "variable") {
      throw new Error(`impedanceMode must be 'fixed' or 'variable'; got '${String(cfg.impedanceMode)}'`);
    }
    if (cfg.nullspaceControl !== "none" && cfg.nullspaceControl !== "position") {
      throw new Error(`nullspaceControl must be 'none' or 'position'; got '${String(cfg.nullspaceControl)}'`);
    }
    if (cfg.motionStiffnessTask.length !== 6) {
      throw new Error(`motionStiffnessTask must be 6-vec; got ${cfg.motionStiffnessTask.length}`);
    }
    if (cfg.motionDampingRatioTask.length !== 6) {
      throw new Error(`motionDampingRatioTask must be 6-vec; got ${cfg.motionDampingRatioTask.length}`);
    }
    this.cfg = cfg;
    this.numEnvs = numEnvs;
    this.numDof = numDof;
    this._target = Array.from({ length: numEnvs }, () => [0, 0, 0, 0, 0, 0, 1]);
    this._variableStiffness = new Array<number[] | null>(numEnvs).fill(null);
  }

  /** Length of one command vector.
   * pose_abs=7, pose_rel=6, wrench_abs=6, force_abs/torque_abs=3.
   * impedanceMode="variable" adds 12 (6 stiffness + 6 damping ratio).
   */
  get actionDim(): number {
    let dim = 0;
    for (const tt of this.cfg.targetTypes) {
      if (tt === "pose_abs") dim += 7;
      else if (tt === "pose_rel") dim += 6;
      else if (tt === "wrench_abs") dim += 6;
      else if (tt === "force_abs" || tt === "torque_abs") dim += 3;
    }
    if (this.cfg.impedanceMode === "variable") dim += 12;
    return dim;
  }

  reset(envIds?: readonly number[]): void {
    const ids = envIds ?? Array.from({ length: this.numEnvs }, (_, i) => i);
    for (const i of ids) {
      this._target[i] = [0, 0, 0, 0, 0, 0, 1];
      this._variableStiffness[i] = null;
    }
  }

  /** Set the target. For pose_abs: command = [px,py,pz,qx,qy,qz,qw].
   * For pose_rel: command = [Δx,Δy,Δz,Δax,Δay,Δaz] axis-angle and
   * eePos/eeQuat MUST be supplied.
   * In `variable` mode, the LAST 12 elements are stiffness (6) + damping (6).
   */
  setCommand(
    command: readonly number[],
    opts: { eePos?: Vec3; eeQuat?: Quat; envIdx?: number } = {},
  ): void {
    const { eePos, eeQuat, envIdx = 0 } = opts;
    if (envIdx < 0 || envIdx >= this.numEnvs) {
      throw new RangeError(`envIdx=${envIdx} out of range [0, ${this.numEnvs})`);
    }
    const expected = this.actionDim;
    if (command.length !== expected) {
      throw new Error(
        `command must be length ${expected} (targetTypes=${JSON.stringify(this.cfg.targetTypes)}, ` +
          `impedanceMode='${this.cfg.impedanceMode}'); got ${command.length}`,
      );
    }
    let cmd: readonly number[] = command;
    if (this.cfg.impedanceMode === "variable") {
      const gainsStart = expected - 12;
      this._variableStiffness[envIdx] = [...command.slice(gainsStart)];
      cmd = command.slice(0, gainsStart);
    }
    let idx = 0;
    for (const tt of this.cfg.targetTypes) {
      if (tt === "pose_abs") {
        this._target[envIdx] = [
          cmd[idx], cmd[idx + 1], cmd[idx + 2],
          cmd[idx + 3], cmd[idx + 4], cmd[idx + 5], cmd[idx + 6],
        ];
        idx += 7;
      } else if (tt === "pose_rel") {
        if (eePos === undefined || eeQuat === undefined) {
          throw new Error("pose_rel requires eePos + eeQuat");
        }
        const delta: number[] = [cmd[idx], cmd[idx + 1], cmd[idx + 2], cmd[idx + 3], cmd[idx + 4], cmd[idx + 5]];
        this._target[envIdx][0] = eePos[0] + delta[0];
        this._target[envIdx][1] = eePos[1] + delta[1];
        this._target[envIdx][2] = eePos[2] + delta[2];
        const angle = Math.sqrt(delta[3] ** 2 + delta[4] ** 2 + delta[5] ** 2);
        let qDelta: Quat;
        if (angle < 1e-9) {
          qDelta = [0, 0, 0, 1];
        } else {
          const ax: Vec3 = [delta[3] / angle, delta[4] / angle, delta[5] / angle];
          const h = angle * 0.5;
          const s = Math.sin(h);
          qDelta = [ax[0] * s, ax[1] * s, ax[2] * s, Math.cos(h)];
        }
        const tq = quatMul(qDelta, eeQuat);
        this._target[envIdx][3] = tq[0];
        this._target[envIdx][4] = tq[1];
        this._target[envIdx][5] = tq[2];
        this._target[envIdx][6] = tq[3];
        idx += 6;
      } else if (tt === "wrench_abs") {
        idx += 6;
      } else {
        idx += 3;
      }
    }
  }

  getTarget(envIdx: number = 0): number[] {
    return [...this._target[envIdx]];
  }

  /** Compute joint torques.
   *
   * Pipeline:
   *   1. pose error e = [pos_err (3-vec); ori_err (axis-angle 3-vec)]
   *   2. velocity error de = -[ee_lin_vel; ee_ang_vel] (target rest = 0)
   *   3. F_task = K_p · e + K_d · de  (K_d = 2 · ζ · sqrt(K_p))
   *   4. τ = Jᵀ · F_task
   *   5. (optional) τ_ns = (I - Jᵀ J⁺ᵀ) · (K_p_ns · (q_target - q) - K_d_ns · qdot)
   *   6. (optional) τ += G(q)
   */
  compute(args: {
    eePos: Vec3;
    eeQuat: Quat;
    eeLinVel: Vec3;
    eeAngVel: Vec3;
    jacobian: number[][];
    jointPos?: readonly number[];
    jointVel?: readonly number[];
    massMatrix?: number[][];
    gravityTorque?: readonly number[];
    nullspaceTargetPos?: readonly number[];
    envIdx?: number;
  }): number[] {
    const {
      eePos, eeQuat, eeLinVel, eeAngVel, jacobian,
      jointPos, jointVel, gravityTorque, nullspaceTargetPos, envIdx = 0,
    } = args;
    if (jacobian.length !== 6) {
      throw new Error(`jacobian must have 6 rows; got ${jacobian.length}`);
    }
    const n = jacobian[0].length;
    if (n !== this.numDof) {
      throw new Error(`jacobian width ${n} != cfg.numDof ${this.numDof}`);
    }
    const target = this._target[envIdx];
    const tPos: Vec3 = [target[0], target[1], target[2]];
    const tQuat: Quat = [target[3], target[4], target[5], target[6]];

    // 1. Pose error
    const posErr: Vec3 = [tPos[0] - eePos[0], tPos[1] - eePos[1], tPos[2] - eePos[2]];
    const qErr = quatMul(tQuat, quatInverse(eeQuat));
    const oriErr = axisAngleVec(qErr);
    const error = [posErr[0], posErr[1], posErr[2], oriErr[0], oriErr[1], oriErr[2]];

    // 2. Velocity error
    const dError = [-eeLinVel[0], -eeLinVel[1], -eeLinVel[2], -eeAngVel[0], -eeAngVel[1], -eeAngVel[2]];

    // 3. Task-space wrench
    let kpTask: readonly number[];
    let kdRatio: readonly number[];
    const vs = this._variableStiffness[envIdx];
    if (this.cfg.impedanceMode === "variable" && vs !== null) {
      kpTask = vs.slice(0, 6);
      kdRatio = vs.slice(6, 12);
    } else {
      kpTask = this.cfg.motionStiffnessTask;
      kdRatio = this.cfg.motionDampingRatioTask;
    }
    const kdTask: number[] = [];
    for (let i = 0; i < 6; i++) {
      kdTask.push(2 * kdRatio[i] * Math.sqrt(Math.max(0, kpTask[i])));
    }
    const FTask: number[] = [];
    for (let i = 0; i < 6; i++) {
      FTask.push(kpTask[i] * error[i] + kdTask[i] * dError[i]);
    }

    // 4. Joint torque via Jᵀ F_task
    const tau = new Array<number>(n).fill(0);
    for (let i = 0; i < n; i++) {
      let s = 0;
      for (let k = 0; k < 6; k++) s += jacobian[k][i] * FTask[k];
      tau[i] = s;
    }

    // 5. Null-space joint regularization
    if (
      this.cfg.nullspaceControl === "position" &&
      nullspaceTargetPos !== undefined &&
      jointPos !== undefined &&
      jointVel !== undefined &&
      n > 6
    ) {
      const kpNs = this.cfg.nullspaceStiffness;
      const kdNs = 2 * this.cfg.nullspaceDampingRatio * Math.sqrt(Math.max(0, kpNs));
      const tauNs: number[] = [];
      for (let i = 0; i < n; i++) {
        tauNs.push(kpNs * (nullspaceTargetPos[i] - jointPos[i]) - kdNs * jointVel[i]);
      }
      const tauNsProj = projectToNullspace(jacobian, tauNs, n, 0.05);
      for (let i = 0; i < n; i++) tau[i] += tauNsProj[i];
    }

    // 6. Gravity compensation
    if (this.cfg.gravityCompensation && gravityTorque !== undefined) {
      const m = Math.min(n, gravityTorque.length);
      for (let i = 0; i < m; i++) tau[i] += gravityTorque[i];
    }
    return tau;
  }
}

// ── Null-space projection (damped pseudoinverse) ─────────────────────────

/**
 * Project tauNs into the null-space of Jacobian J via:
 *   P_null = I - Jᵀ (J Jᵀ + λ²I)⁻¹ J
 *   tau_proj = P_null · tauNs
 *
 * Uses Gauss-Jordan 6×6 inverse with partial pivoting + 1e-18 singularity
 * guard. Damping (λ²=0.0025) gives numerical stability near singular
 * configurations.
 */
function projectToNullspace(
  J: readonly number[][],
  tauNs: readonly number[],
  n: number,
  lam: number,
): number[] {
  // Step 1: J · tauNs (6-vec)
  const Jtau = new Array<number>(6).fill(0);
  for (let k = 0; k < 6; k++) {
    let s = 0;
    for (let i = 0; i < n; i++) s += J[k][i] * tauNs[i];
    Jtau[k] = s;
  }
  // Step 2: solve (J Jᵀ + λ²I) y = Jtau
  const lam2 = lam * lam;
  const A: number[][] = [];
  for (let i = 0; i < 6; i++) {
    const row = new Array<number>(6).fill(0);
    for (let j = 0; j < 6; j++) {
      let s = 0;
      for (let k = 0; k < n; k++) s += J[i][k] * J[j][k];
      row[j] = s + (i === j ? lam2 : 0);
    }
    A.push(row);
  }
  const aug: number[][] = A.map((r, i) => [...r, Jtau[i]]);
  for (let col = 0; col < 6; col++) {
    let piv = col;
    let maxAbs = Math.abs(aug[col][col]);
    for (let r = col + 1; r < 6; r++) {
      if (Math.abs(aug[r][col]) > maxAbs) {
        maxAbs = Math.abs(aug[r][col]);
        piv = r;
      }
    }
    if (maxAbs < 1e-18) continue;
    if (piv !== col) {
      const tmp = aug[col];
      aug[col] = aug[piv];
      aug[piv] = tmp;
    }
    const pv = aug[col][col];
    for (let j = col; j < 7; j++) aug[col][j] /= pv;
    for (let r = 0; r < 6; r++) {
      if (r === col) continue;
      const f = aug[r][col];
      if (Math.abs(f) < 1e-18) continue;
      for (let j = col; j < 7; j++) aug[r][j] -= f * aug[col][j];
    }
  }
  const y = aug.map((r) => r[6]);
  // Step 3: Jᵀ · y  (n-vec)
  const Jty = new Array<number>(n).fill(0);
  for (let i = 0; i < n; i++) {
    let s = 0;
    for (let k = 0; k < 6; k++) s += J[k][i] * y[k];
    Jty[i] = s;
  }
  // Step 4: tau_proj = tauNs - Jty
  const out = new Array<number>(n);
  for (let i = 0; i < n; i++) out[i] = tauNs[i] - Jty[i];
  return out;
}
